qutebrowser/misc/sql.py

```diff
@@ -0,0 +1,166 @@
+# vim: ft=python fileencoding=utf-8 sts=4 sw=4 et:
+
+# Copyright 2016 Ryan Roden-Corrent (rcorre) <ryan@rcorre.net>
+#
+# This file is part of qutebrowser.
+#
+# qutebrowser is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# qutebrowser is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with qutebrowser.  If not, see <http://www.gnu.org/licenses/>.
+
+"""Provides access to an in-memory sqlite database."""
+
+from PyQt5.QtCore import QObject, pyqtSignal
+from PyQt5.QtSql import QSqlDatabase, QSqlQuery
+
+from qutebrowser.utils import log
+
+
+def init():
+    """Initialize the SQL database connection."""
+    database = QSqlDatabase.addDatabase('QSQLITE')
+    # In-memory database, see https://sqlite.org/inmemorydb.html
+    database.setDatabaseName(':memory:')
+    if not database.open():
+        raise SqlException("Failed to open in-memory sqlite database")
+
+
+def close():
+    """Close the SQL connection."""
+    QSqlDatabase.removeDatabase(QSqlDatabase.database().connectionName())
+
+
+def _run_query(querystr, *values):
+    """Run the given SQL query string on the database.
+
+    Args:
+        values: positional parameter bindings.
+    """
+    log.completion.debug('Running SQL query: "{}"'.format(querystr))
+    database = QSqlDatabase.database()
+    query = QSqlQuery(database)
+    query.prepare(querystr)
+    for val in values:
+        query.addBindValue(val)
+    log.completion.debug('Query bindings: {}'.format(query.boundValues()))
+    if not query.exec_():
+        raise SqlException('Failed to exec query "{}": "{}"'.format(
+                           querystr, query.lastError().text()))
+    return query
+
+
+class SqlTable(QObject):
```

Comment:
Thanks! That actually answers your question about `KeyError` though. I saw this as a dict-like interface over a sql table. When you try to access or delete a key that doesn't exist, you get a `KeyError` just as you would if it were a python dict. I didn't see a reason to invent a new exception just for that (I do have `SqlException`, but that's more for things like 'your SQL connection died')